### **Autor**: David Roca Tauste

---

<div style="text-align: center;">
    <h1>UNIDAD DIDÁCTICA 4</h1>
    <h1>ANÁLISIS Y BÚSQUEDA DE RESPUESTAS EN DATOS</h1>
    <h3>Curso: Curso de especialización de inteligencia artificial y Big Data</h3>
    <h3>Módulo: Sistemas de Big Data – Curso: IABD</h3>
</div>

---

# Enunciado del Proyecto en R: Análisis de Ventas y Gastos en una Empresa con 5 Tiendas

## Descripción
Una empresa cuenta con **5 tiendas**, y se desea analizar sus ventas y gastos mensuales durante **un año (12 meses)**. Los gastos están divididos en las siguientes categorías:
- **Material**
- **Personal**
- **Publicidad**
- **Gastos eléctricos**

Deberéis desarrollar un programa en **R** que permita analizar los datos generados de manera lógica (sin valores extremadamente dispares).

---

<h2><b><u>Objetivos del Programa</u></b></h2>

<h2><b><u>0. Preparación del análisis</u></b></h2>

En este apartado se va a mostrar el código necesario para poder realizar los análisis solicitados.

In [6]:
# Librerías utilizadas
import numpy as np
import pandas as pd
from scipy.stats import mode

In [7]:
# Configurar semilla para que los datos sean siempre los mismos
np.random.seed(123)

# Definir los nombres de las tiendas y los meses
tiendas = ["Tienda Sevilla", "Tienda Barcelona", "Tienda Valencia", "Tienda Baleares", "Tienda Santiago"]
meses = ["Enero", "Febrero", "Marzo", "Abril", "Mayo", "Junio", "Julio", "Agosto", "Septiembre", "Octubre", "Noviembre", "Diciembre"]

# Generar datos aleatorios dentro de los rangos especificados
cantidad_datos_necesarios = len(meses) * len(tiendas)

ingresos = np.random.randint(50000, 150001, size=(len(tiendas), len(meses)))
gastos_material = np.random.randint(5000, 20001, size=(len(tiendas), len(meses)))
gastos_personal = np.random.randint(20000, 50001, size=(len(tiendas), len(meses)))
gastos_publicidad = np.random.randint(3000, 10001, size=(len(tiendas), len(meses)))
gastos_electricos = np.random.randint(2000, 7001, size=(len(tiendas), len(meses)))

total_gastos = gastos_material + gastos_personal + gastos_publicidad + gastos_electricos
beneficio = ingresos - total_gastos

# Convertir a DataFrames
df_ingresos = pd.DataFrame(ingresos, index=tiendas, columns=meses)
df_gastos = pd.DataFrame(total_gastos, index=tiendas, columns=meses)
df_beneficio = pd.DataFrame(beneficio, index=tiendas, columns=meses)

**Explicación:**

Este código genera datos aleatorios para simular los ingresos, gastos y beneficios de varias tiendas a lo largo de un año. A continuación, se explica de manera más detallada:

- **Configurar la semilla aleatoria "np.random.seed(123)"**: para garantizar que los valores aleatorios generados sean siempre los mismos en cada ejecución.

- **Definir los nombres de tiendas y meses**:
    - Tiendas: lista que contiene los nombres de cinco tiendas.
    - Meses: lista que contiene los nombres de los doce meses del año.

- **Calcular la cantidad de datos necesarios**: se multiplica la cantidad de meses "len(meses)" por la cantidad de tiendas "len(tiendas)" para determinar cuántos valores aleatorios se necesitan.

- **Generación de datos aleatorios**<br> 
Se crean arrays de numpy de valores aleatorios utilizando "np.random.randint" con dimensiones de 5 filas (tiendas) por 12 columnas (meses), con los siguientes rangos:
    - ingresos: entre 50.000 y 150.000.
    - gastos_material: entre 5.000 y 20.000.
    - gastos_personal: entre 20.000 y 50.000.
    - gastos_publicidad: entre 3.000 y 10.000.
    - gastos_electricos: entre 2.000 y 7.000.

- **Cálculo de gastos totales y beneficios**
    - total_gastos: suma de todos los gastos de cada tienda en cada mes.
    - beneficio: diferencia entre los ingresos y los gastos totales.

- **Convertir a DataFrames**<br>
Para facilitar el análisis, los datos generados se convierten en DataFrames de Pandas, donde:
    - Filas: representan las tiendas.
    - Columnas: representan los meses.

---

<h2><b><u>1. Cálculo de Totales</u></b></h2>

**A nivel global** (sumando todas las tiendas):
- Total de ingresos mensuales.
- Total de cada tipo de gasto mensual.
- Beneficio mensual total (beneficio = ingresos - gastos).

**A nivel de cada tienda**:
- Ingresos mensuales de cada tienda.
- Total de cada tipo de gasto mensual por tienda.
- Beneficio mensual por tienda.

In [8]:
# 1. Cálculo de totales globales y de cada tienda
print("1. CALCULO DE TOTALES:")
print("Ingresos totales por mes:")
print(df_ingresos.sum())
print("Gastos totales por mes:")
print(df_gastos.sum())
print("Beneficio total por mes:")
print(df_beneficio.sum())

print("\nIngresos totales por tienda:")
print(df_ingresos.sum(axis=1))
print("Gastos totales por tienda:")
print(df_gastos.sum(axis=1))
print("Beneficio total por tienda:")
print(df_beneficio.sum(axis=1))

1. CALCULO DE TOTALES:
Ingresos totales por mes:
Enero         407837
Febrero       524909
Marzo         476165
Abril         525388
Mayo          448875
Junio         545977
Julio         524812
Agosto        633983
Septiembre    436120
Octubre       400953
Noviembre     525937
Diciembre     445280
dtype: int64
Gastos totales por mes:
Enero         296493
Febrero       316296
Marzo         281444
Abril         340453
Mayo          322036
Junio         321858
Julio         286209
Agosto        306727
Septiembre    282468
Octubre       298460
Noviembre     287193
Diciembre     298207
dtype: int64
Beneficio total por mes:
Enero         111344
Febrero       208613
Marzo         194721
Abril         184935
Mayo          126839
Junio         224119
Julio         238603
Agosto        327256
Septiembre    153652
Octubre       102493
Noviembre     238744
Diciembre     147073
dtype: int64

Ingresos totales por tienda:
Tienda Sevilla      1153558
Tienda Barcelona    1229653
Tienda Valencia     1

**Explicación**:
- `sum()`: suma las filas de un dataframe.
- `sum(axis=1)`: suma las columnas de un dataframe.

---

<h2><b><u>2. Análisis de los Meses Clave</u></b></h2>

**A nivel global**:
- Mes con mayores ingresos y mes con menores ingresos.
- Mes con mayor beneficio y mes con menor beneficio.

**A nivel de cada tienda**:
- Mes con mayores ingresos y mes con menores ingresos por tienda.
- Mes con mayor beneficio y mes con menor beneficio por tienda.

In [9]:
# 2. Análisis de los meses clave
print("2. ANALISIS DE LOS MESES CLAVE:")
print("Mes con mayores ingresos:", df_ingresos.sum().idxmax())
print("Mes con menores ingresos:", df_ingresos.sum().idxmin())
print("Mes con mayor beneficio:", df_beneficio.sum().idxmax())
print("Mes con menor beneficio:", df_beneficio.sum().idxmin())

print("\nMes con mayores ingresos por tienda:")
print(df_ingresos.idxmax(axis=1))
print("Mes con menores ingresos por tienda:")
print(df_ingresos.idxmin(axis=1))
print("Mes con mayor beneficio por tienda:")
print(df_beneficio.idxmax(axis=1))
print("Mes con menor beneficio por tienda:")
print(df_beneficio.idxmin(axis=1))

2. ANALISIS DE LOS MESES CLAVE:
Mes con mayores ingresos: Agosto
Mes con menores ingresos: Octubre
Mes con mayor beneficio: Agosto
Mes con menor beneficio: Octubre

Mes con mayores ingresos por tienda:
Tienda Sevilla      Septiembre
Tienda Barcelona        Agosto
Tienda Valencia          Junio
Tienda Baleares           Mayo
Tienda Santiago          Junio
dtype: object
Mes con menores ingresos por tienda:
Tienda Sevilla           Junio
Tienda Barcelona    Septiembre
Tienda Valencia        Octubre
Tienda Baleares          Marzo
Tienda Santiago        Febrero
dtype: object
Mes con mayor beneficio por tienda:
Tienda Sevilla      Septiembre
Tienda Barcelona        Agosto
Tienda Valencia          Marzo
Tienda Baleares           Mayo
Tienda Santiago          Junio
dtype: object
Mes con menor beneficio por tienda:
Tienda Sevilla          Marzo
Tienda Barcelona         Mayo
Tienda Valencia     Noviembre
Tienda Baleares       Octubre
Tienda Santiago         Abril
dtype: object


**Explicación**
- `sum().idxmax()`: "sum()" suma las filas del dataframe e "idxmax()" obtiene el nombre de la columna con el valor máximo.
- `sum().idxmin()`: "sum()" suma las filas del dataframe e "idxmin()" obtiene el nombre de la columna con el valor mínimo.

- `idxmax(axis=1)`: obtiene el nombre de la columna, de la fila con el valor máximo.
- `idxmin(axis=1)`: obtiene el nombre de la columna, de la fila con el valor mínimo.

---

<h2><b><u>3. Estadísticas Mensuales</u></b></h2>

**A nivel global**:
- Media, mediana, moda y desviación típica de los ingresos mensuales.
- Media, mediana, moda y desviación típica del beneficio mensual.

**A nivel de cada tienda**:
- Media, mediana, moda y desviación típica de los ingresos por tienda.
- Media, mediana, moda y desviación típica del beneficio por tienda.

In [10]:
# 3. Estadísticas mensuales
print("3. ESTADISTICAS MENSUALES:")
# A nivel global
print("Media de los ingresos mensuales:", df_ingresos.sum().mean())
print("Mediana de los ingresos mensuales:", df_ingresos.sum().median())
moda_ingresos = mode(df_ingresos, axis=None).mode
print("Moda de los ingresos mensuales: ", moda_ingresos)
print("Desviacion tipica de los ingresos mensuales:", df_ingresos.sum().std())

print("\nMedia del beneficio mensual:", df_beneficio.sum().mean())
print("Mediana del beneficio mensual:", df_beneficio.sum().median())
moda_beneficio = mode(df_beneficio, axis=None).mode
print("Moda del beneficio mensual: ", moda_beneficio)
print("Desviacion tipica del beneficio mensual:", df_beneficio.sum().std())

# A nivel de cada tienda
print("\nMedia de los ingresos mensuales por tienda:")
print(df_ingresos.mean(axis=1))
print("Mediana de los ingresos mensuales por tienda:")
print(df_ingresos.median(axis=1))
moda_ingresos_tienda = mode(df_ingresos, axis=1).mode
print("Moda de los ingresos mensuales por tienda: ", moda_ingresos_tienda)
print("Desviacion tipica de los ingresos mensuales por tienda:")
print(df_ingresos.std(axis=1))

print("\nMedia del beneficio mensual por tienda:")
print(df_beneficio.mean(axis=1))
print("Mediana del beneficio mensual por tienda:")
print(df_beneficio.median(axis=1))
moda_beneficio_tienda = mode(df_beneficio, axis=1).mode
print("Moda del beneficio mensual por tienda: ", moda_beneficio_tienda)
print("Desviacion tipica del beneficio mensual por tienda:")
print(df_beneficio.std(axis=1))

3. ESTADISTICAS MENSUALES:
Media de los ingresos mensuales: 491353.0
Mediana de los ingresos mensuales: 500488.5
Moda de los ingresos mensuales:  52428
Desviacion tipica de los ingresos mensuales: 67610.23730713621

Media del beneficio mensual: 188199.33333333334
Mediana del beneficio mensual: 189828.0
Moda del beneficio mensual:  -8679
Desviacion tipica del beneficio mensual: 64733.35490559201

Media de los ingresos mensuales por tienda:
Tienda Sevilla       96129.833333
Tienda Barcelona    102471.083333
Tienda Valencia      85960.166667
Tienda Baleares     107329.666667
Tienda Santiago      99462.250000
dtype: float64
Mediana de los ingresos mensuales por tienda:
Tienda Sevilla      100806.0
Tienda Barcelona    105577.0
Tienda Valencia      75177.0
Tienda Baleares     117426.0
Tienda Santiago     105491.5
dtype: float64
Moda de los ingresos mensuales por tienda:  [65377 64944 52428 61894 59610]
Desviacion tipica de los ingresos mensuales por tienda:
Tienda Sevilla      25527.809303
T

**Explicación**:
- `mean()`: calcula la media del dataframe.
- `median()`: calcula la mediana del dataframe.
- `mode(df_ingresos, axis=None).mode`: obtiene la moda del dataframe "df_ingresos". Para ello, se utiliza una librería de Scikit-learn.
- `std()`: calcula la desviación estándar del dataframe.

---

<h2><b><u>4. Conclusiones</u></b></h2>

<h3><b>1. Conclusiones extraídas de cada análisis</h3>

Cálculo de Totales (Ingresos, Gastos y Beneficio)</b>
- Los ingresos totales mensuales varían entre 400.953€ (octubre, el menor) y 633.983€ (agosto, el mayor).
- Los gastos mensuales varían entre 281.444€ (marzo, el menor) y 340.453€ (abril, el mayor), lo que indica fluctuaciones en los gastos.
- El beneficio total por mes varía entre 102.493€ (octubre, el menor) y 327.256€ (agosto, el mayor), por lo tanto, octubre es el mes con peor rendimiento y agosto el de mejor rendimiento.

<b>Análisis de Meses Clave</b>
- Agosto es el mes con mayores ingresos y beneficios, lo que indica un aumento de la demanda en este mes.
- Octubre es el mes con menores ingresos y beneficios.
- A nivel de tienda, cada una presenta su propio mes de mayor y menor rendimiento, lo que puede estar relacionado con la ubicación o con estrategias comerciales distintas.

<b>Estadísticas Mensuales</b>
- La media de ingresos mensuales es de 491.353€, con una mediana de 500.488,5€ y una moda de 52428€. La desviación estándar es de 67.610,24€, lo que indica estabilidad en los ingresos.
- El beneficio mensual tiene una media de 188.199,33€, una mediana de 189.828€ y una moda de -8679€. En este caso, la desviación estándar es de 64.733,35€, lo que indica que los beneficios son más variables.

<h3><b>2. Comparación de los resultados entre tiendas y determinación de tiendas con el mejor rendimiento</h3>

Ingresos totales</b>
- La tienda con mayores ingresos anuales es Baleares (1.287.956€), seguida de Barcelona (1.229.653€).
- La tienda con menores ingresos es Valencia (1.031.522€).

<b>Beneficio total</b>
- La tienda con mayor beneficio es Baleares (530.401€), seguida de Barcelona (528.267€).
- La tienda con menor beneficio es Valencia (299.511€), lo que sugiere que, aunque genera ingresos, sus gastos son altos.

<b>Desviación estándar y estabilidad</b>
- Baleares y Barcelona tienen los ingresos y beneficios más altos con desviaciones estándar bajas, lo que indica estabilidad en su rendimiento.
- Valencia tiene una desviación estándar más alta, lo que indica una menor estabilidad en sus beneficios e ingresos.

<h3><b>3. Identificación de patrones y tendencias</h3>

Tendencias globales</b>
- Estacionalidad: agosto es el mes con mayor rendimiento y octubre el mes con menor rendimiento.
- Fluctuaciones de gastos: los gastos no son constantes y pueden afectar a la rentabilidad mensual de las tiendas.

<b>Patrones por tienda</b>
- Las tiendas con mejor rendimiento (Baleares y Barcelona) muestran una mayor estabilidad en sus ingresos y márgenes de beneficio más altos.
- Las tiendas con mayor variabilidad en ingresos y beneficios (Valencia y Sevilla) podrían tener una demanda más inestable o sufrir gastos operativos más elevados.
- Valencia tiene los ingresos y el beneficio más bajos, lo que indica que su margen de rentabilidad podría mejorar, ya sea optimizando los gastos o ajustando sus estrategias de venta.